Scripts/Proxies/ProxyAnimator.cs

```diff
@@ -0,0 +1,136 @@
+﻿using System.Collections.Generic;
+using UnityEditor.Experimental.EditorVR;
+using UnityEditor.Experimental.EditorVR.Proxies;
+using UnityEngine;
+using UnityEngine.InputNew;
+
+[ProcessInput(1)]
+[RequireComponent(typeof(ProxyHelper))]
+public class ProxyAnimator : MonoBehaviour, ICustomActionMap
+{
+	class TransformInfo
+	{
+		public Vector3 initialPosition;
+		public Vector3 initialRotation;
+		public Vector3 positionOffset;
+		public Vector3 rotationOffset;
+
+		public void Reset()
+		{
+			positionOffset = Vector3.zero;
+			rotationOffset = Vector3.zero;
+		}
+	}
+
+	[SerializeField]
+	ActionMap m_ProxyActionMap;
+
+	ProxyHelper.ButtonObject[] m_Buttons;
+	InputControl[] m_Controls;
+
+	readonly Dictionary<Transform, TransformInfo> m_TransformInfos = new Dictionary<Transform, TransformInfo>();
+
+	public ActionMap actionMap { get { return m_ProxyActionMap; } }
+	public bool ignoreLocking { get { return true; } }
+
+	void Start()
+	{
+		m_Buttons = GetComponent<ProxyHelper>().buttons;
+	}
+
+	public void ProcessInput(ActionMapInput input, ConsumeControlDelegate consumeControl)
+	{
+		if (m_Buttons == null)
+			return;
+
+		var length = m_Buttons.Length;
+		if (m_Controls == null)
```

Comment:
Is there some kind of order-of-operations thing where this has to happen in ProcessInput?
Should there be some kind of higher level "InitializeInput" function that IProcessInput implementors need to handle?

Reply:
This isn't a bad idea. The reason it has to happen here is because this is (by design) the only place where we have access to the ActionMapInput which was created from our ActionMap. The idea was to make it awkward for people to use input in an Update method because then we can't control the order in which input is processed.

That said, there are many other places where we break this convention, for example ILinkedObject tools that create a private field to store the ActionMapInput so that the shared updater can access input from linked tools. They would also benefit from returning to the original implementation of ICustomActionMap, which included an ActionMapInput property with a public setter that the system would use to give back the AMI that it creates when it spawns the tool.